torch/csrc/Module.cpp

```diff
@@ -688,6 +690,50 @@ PyObject* THPModule_isEnabledXNNPACK(PyObject* _unused, PyObject* noargs) {
     Py_RETURN_FALSE;
 }
 
+PyObject* THPModule_willEngineExecuteNode(PyObject* _unused, PyObject* arg) {
+  HANDLE_TH_ERRORS
+  bool isTHPFunction = THPFunction_Check(arg);
+  bool isTHPCppFunction = torch::autograd::THPCppFunction_Check(arg);
+  THPUtils_assert(
+      isTHPFunction || isTHPCppFunction,
+      "_will_engine_execute_node expects an grad_fn, "
+      "but got %s",
+      THPUtils_typename(arg));
+  const auto exec_info = torch::autograd::get_current_graph_task_exec_info();
+  THPUtils_assert(
+      exec_info,
+      "_get_should_execute_nodes should only be called during the backward pass");
+  torch::autograd::Node* node;
+  std::shared_ptr<torch::autograd::Node> node_sp;
+  if (isTHPFunction) {
+    node_sp = ((THPFunction*)arg)->cdata.lock();
+    node = node_sp.get();
+  } else {
+    node = ((torch::autograd::THPCppFunction*)arg)->cdata.get();
+  }
+  if (exec_info->empty()) {
+    // .backward() without inputs= arg
+    const auto nodes_in_graph =
+        torch::autograd::get_current_graph_task_nodes_in_graph();
+    auto it = nodes_in_graph->find(node);
+    if (it == nodes_in_graph->end()) {
+      Py_RETURN_FALSE;
+    } else {
+      Py_RETURN_TRUE;
+    }
+  } else {
```

Comment:
Why do we need this case? The new field covers this case as well.

Reply:
We should need both because compute_dependencies just traverses through the graph without taking into account the inputs= argument.

Reply:
No it does not? Otherwise the `dependencies_` counter would be wrong?

Reply:
From reading the code again, what happens later is that we want to use dependencies_ to see if a given node is ready to be executed, but first we use exec_info to decide whether we want to execute the node in the first place, so it is indeed possible that a node will have an entry in dependencies_, but won't actually be executed.

Reply:
Ho right at the border of the of the graph there is one more node for which the hook won't be called?
But is that the expected behavior btw? (there is an issue for that somewhere right?)

Reply:
Ohh that is a very good catch, neither variables hooks registered to leafs nor function hooks registered to acc grad are called when .grad is used. 

As you can see from the test, exec_info_ does not know about that either though. 

I think we should error out in this case, as opposed to silently returning false for accumulate grad nodes. If someone passes in an acc grad node to our function when running .grad that is probably user error.